Add kernel connection API

swlynch99 opened the pull request:
This is an attempt at addressing #2362 by following the API suggested there. 

It introduces a new `ExternalConnection` type which allows users to pass received session tickets back to rustls and perform key updates. An `ExternalConnection` can be constructed from an `Unbuffered(Client|Server)Connection` by calling `dangerous_into_external_connection` which will return both an `ExtractedSecrets` and an `ExternalConnection`. It does not include support any other connection types at this time. However, it should be easy enough to extend in the future as more as new features become needed.

Internally this is implemented by adding a new `ExternalState` trait and a `State::into_external_state` conversion method. `ExternalState` is implemented for the `ExpectTraffic` state for tls 1.2 and 1.3, and for server and client modes, respectively. 

I have also included a complete example that shows how to use `ExternalConnection` with a kTLS connection. It ends up serving more as an example in getting kTLS to work using rustls than actually using `ExternalConnection`, since `ExternalConnection` has a pretty minimal API surface, but I think it works better that way.

This PR ended up being quite a bit larger than I thought it would be initially. If there's anything I can do that would help make this easier to review (e.g. moving the ktls example to a separate PR), please let me know.

Closes #2362

ctz commented:
First, I think the `dangerous_into_external_connection()` API is an excellent way to do this and maybe I'd go further to suggest that this PR could deprecate the `dangerous_extract_secrets()` API, and eviscerate its internal implementation (eg, by making `dangerous_extract_secrets()` just do `let (secrets, _) = self.dangerous_into_external_connection()?; Ok(secrets)`).

I think there is space here to reduce the public API scope of `ExternalConnection` quite a bit. Observing the idea that KTLS users are exchanging complexity for big performance and density improvements, I think that committing to keeping things like the peer's certificate around while servicing the connection is not in keeping with that goal. So I would suggest removing much of that API surface for now -- since the caller needs to ensure the handshake is complete _before_ calling `dangerous_into_external_connection()`, if they need the peer certificate, handshake shape, ALPN, etc, they can obtain this before the conversion. (That doesn't stop us from having to store it internally, but gives us some future leeway in how we do that to improve density.)

djc commented:
Agreed. Also, a bit of bikeshedding: I don't love `External` as it feels a little generic. Was thinking maybe `Traffic` would be a good prefix for this that is a little more meaningful in communicating how it's different?

swlynch99 commented:
I'm going to make changes as new commits for now since it is easier to see what is changing. I'll rework things into a more coherent commit history before this gets merged.

Quick summary of the new changes before I respond to individual comments:
- I have made `ExternalConnection` generic over `Data` so that users with a server-side connection cannot call `handle_new_session_ticket`. This is currently just a facade using `PhantomData` over non-generic internals but it gives freedom in the future if something needs access to `Data`. My main motivation for this, though, is disallowing calls to `handle_new_session_ticket` on server side connections.
- I have removed a bunch of the public API methods from `ExternalConnection`.
- I have reimplemented `dangerous_into_secrets` on top of `dangerous_into_external_connection`.  The one change in behaviour is that it now emits an error if there are messages stored in `sendable_tls` instead of giving unusable secrets.

> I'd go further to suggest that this PR could deprecate the dangerous_extract_secrets() API, and eviscerate its internal implementation.

I have done all of this except the deprecation. I'll add a commit for the deprecation bit once we've decided on the final name for the type.

> I think there is space here to reduce the public API scope of ExternalConnection quite a bit.

I have reduced the API surface down to just the negotiated protocol version and the chosen cipher suite. The protocol version needs to be kept anyways and the cipher suite is the only way to get the confidentiality limit (and is needed for kTLS besides). This has not appreciably reduced the size of `ExternalConnection`, unfortunately, since the quic data is ~600B and everything else is ~80B.

> I don't love External as it feels a little generic. Was thinking maybe Traffic would be a good prefix for this that is a little more meaningful in communicating how it's different?

Up front, I don't have a strong preference here. This is your bikeshed so ultimately you get the final say :)

My thought process behind `ExternalConnection` was something like "it's a connection that's managed externally to rustls" -> `ExternalConnection`. Not sure if that makes it better or worse.

`TrafficConnection` sounds a little off to me. It might be because all the other prefixes are adjectives?. As an alternative, what if we just called it `Connection` like in the quic module? Then you have `external::Connection`, `external::ClientConnection`, and `external::ServerConnection`. I'm not sure if that's better, on second thought, but I'll leave it here as an alternative.

djc commented:
> `TrafficConnection` sounds a little off to me. It might be because all the other prefixes are adjectives?. As an alternative, what if we just called it `Connection` like in the quic module? Then you have `external::Connection`, `external::ClientConnection`, and `external::ServerConnection`. I'm not sure if that's better, on second thought, but I'll leave it here as an alternative.

That's fair. I think my beef with "external" is that it leaves open the question of "external to what"?

swlynch99 commented:
@djc I've been thinking on and off about the naming for the last few days and have not been able to come up with a better name. So I'm putting this back on your side of the court: if you want me to rename type+module to `TrafficConnection` and `traffic` then I'll do so, if you think it is ok staying as `ExternalConnection` + `external`, then I'll leave it as is. I don't have a strong enough opinion on this to stall the PR over it :)

I will say that when I named it my thought process was that it is a connection that is "external to rustls". i.e. that the user has to implement the connection themselves, whereas for the other connection types the user just provides/receives bytes and rustls handles the rest.

I think I have addressed all the points brought up so far. The only thing left is deprecating `dangerous_external_secrets` which I will do after naming is settled. Let me know if there is anything I can do to make the review easier, as I realize this is a big PR. Otherwise, I'll leave this until you guys have bandwidth to review.

ctz commented:
I don't mind the `ExternalConnection` naming, but WDYT about `KernelConnection`?

swlynch99 commented:
I have gone ahead and done the rename. I have also deprecated `dangerous_extract_secrets` in favour of the new `dangerous_into_kernel_connection` method. Finally, I have cleaned up the commit history into something a bit more coherent.

djc commented on rustls/src/conn/kernel.rs:
```diff
@@ -0,0 +1,223 @@
+use crate::server::ServerConnectionData;
+use crate::{CommonState, ConnectionTrafficSecrets, Error, ProtocolVersion, SupportedCipherSuite};
+
+/// A kernel client connection.
+///
+/// The only difference for kernel client vs server connections is that client
+/// connections have support for recording new session tickets.
+pub type KernelClientConnection = KernelConnection<ClientConnectionData>;
+
+/// A kernel server connection.
+///
+/// The only difference for kernel client vs server connections is that client
+/// connections have support for recording new session tickets.
+pub type KernelServerConnection = KernelConnection<ServerConnectionData>;
```
I'd prefer to get rid of these type aliases -- I think these generally don't make code clearer but instead obfuscate it in more indirection.

djc commented on rustls/src/conn/kernel.rs:
```diff
@@ -0,0 +1,223 @@
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub(crate) enum Direction {
+    Transmit,
+    Receive,
+}
```
Nit: move this to the bottom of the module.

djc commented on rustls/src/lib.rs:
```diff
@@ -533,6 +533,72 @@ pub mod unbuffered {
     };
 }
 
+/// Kernel connection API.
```
Nit: I think this docstring should be inside the `conn/kernel.rs` file. Could we just do `pub use conn::kernel;`?

djc commented on rustls/src/server/tls12.rs:
```diff
@@ -998,7 +999,30 @@ impl State<ServerConnectionData> for ExpectTraffic {
+        _cx: &mut ExternalContext<'_>,
+        _message: &NewSessionTicketPayloadTls13,
+    ) -> Result<(), Error> {
+        Err(Error::InappropriateMessage {
```
Don't we have a better error variant for wrong protocol version? If not, maybe we should create one?

swlynch99 replied:
This gets handled higher up in https://github.com/swlynch99/rustls/blob/external-connection/rustls/src/conn/kernel.rs#L226. I'm happy to change the error here, but this is dead code so I just mirrored what rustls returns in this case for a normal connection.

ctz replied:
I don't mind that; maybe we could put a `&'static str` in there so we can work out which one was incorrect, if someone who reports seeing such an error.

swlynch99 replied:
Here's what I have done instead:
- For `server::tls12::ExpectTraffic` I have added a useful error message that matches what `KernelConnection::handle_new_session_ticket` returns
- For `client::tls1[23]::ExpectTraffic` I have put `unreachable!` with a useful error message. I have also added a note that the method is only really supposed to be called for client connections.

All three are dead code, but the client ones should definitely never be called (and I've added a note to the `KernelState` documentation about this) whereas the server-side of the connection should just be straight-up unreachable.

djc commented on rustls/src/server/tls13.rs:
```diff
@@ -1499,11 +1500,38 @@ impl State<ServerConnectionData> for ExpectTraffic {
 
+impl ExternalState for ExpectTraffic {
+    fn update_secrets(&mut self, dir: Direction) -> Result<ConnectionTrafficSecrets, Error> {
+        let side = match dir {
+            Direction::Transmit => Side::Server,
+            Direction::Receive => Side::Client,
+        };
```
Nit: suggest inlining these matches everywhere.

swlynch99 replied:
I'm not sure where you want to inline these? This match statement is different based on whether we are on the server or client side of the connection and we don't have that info in `KernelConnection`

djc replied:
I meant like:

```rust
self.key_schedule.refresh_traffic_secret(match dir {
    ..
})
```

djc commented on rustls/src/tls13/key_schedule.rs:
```diff
@@ -534,26 +534,31 @@ impl KeyScheduleTraffic {
+            self.ks.suite.hkdf_provider,
+            self.ks.suite.aead_alg.key_len(),
+        );
+        let secrets = self
```
Nit: no need to bind `secrets`, just inline it.

djc commented on rustls/src/tls13/key_schedule.rs:
```diff
@@ -577,6 +582,15 @@ impl KeyScheduleTraffic {
     }
 }
 
+fn expand_secret(secret: &OkmBlock, hkdf: &'static dyn Hkdf, aead_key_len: usize) -> (AeadKey, Iv) {
```
Nit: please move the extraction of this function into a separate commit.

djc commented on rustls/tests/runners/kernel.rs:
```diff
@@ -0,0 +1,22 @@
+#![allow(clippy::duplicate_mod)]
+
+#[macro_use]
+mod macros;
+
+#[cfg(feature = "ring")]
+#[path = "."]
+mod tests_with_ring {
+    provider_ring!();
+
+    #[path = "../kernel.rs"]
+    mod kernel;
+}
+
+#[cfg(feature = "aws_lc_rs")]
+#[path = "."]
+mod tests_with_aws_lc_rs {
+    provider_aws_lc_rs!();
+
+    #[path = "../kernel.rs"]
+    mod tests;
+}
```
Can we do this without adding yet another small file?

swlynch99 replied:
Sure. Where should I put the tests? I can stick them under `unbuffered`

djc replied:
Yup, that sounds okay.

djc commented on examples/src/bin/ktls-client.rs:
```diff
@@ -0,0 +1,666 @@
+//! This is an example showing how to use rustls' external connection API in
```
I'm not sure we want to maintain all this stuff in this crate...

swlynch99 replied:
That's reasonable. This is really the example I wished I had when I was trying to figure out kTLS. I'm fine if it lives elsewhere as long as I can link it in the docs.

If you would rather not maintain it then let me know and I'll move it to its own repo.

ctz replied:
I think we should use this example as inspiration when integrating this new API into the ktls crate, but probably not keeping it in this repo.

swlynch99 replied:
I have removed the example. Depending on how things go I may follow up later with a link to either a separate example repo or the ktls crate in a follow-up PR.

djc commented on rustls/src/client/client_conn.rs:
```diff
@@ -851,6 +851,8 @@ impl UnbufferedClientConnection {
 
     /// Extract secrets, so they can be used when configuring kTLS, for example.
     /// Should be used with care as it exposes secret key material.
+    #[deprecated = "dangerous_extract_secrets does not support session tickets or \
```
Nit: suffix function/method names with `()` to make their usage more obvious.

djc commented:
I think this is generally the right direction. I'm on the fence about the example code -- while it's nice to have a worked example, it seems to spend a lot of code on the details of the kernel API, while the API surface we have to maintain in order to make it work appears to be substantially smaller.

swlynch99 commented on rustls/src/conn/kernel.rs:
```diff
@@ -0,0 +1,223 @@
+pub struct KernelConnection<Data> {
+    state: Box<dyn ExternalState>,
+
+    peer_certificates: Option<CertificateChain<'static>>,
+    quic: Quic,
+
+    negotiated_version: ProtocolVersion,
+    protocol: Protocol,
+    suite: SupportedCipherSuite,
```
- `peer_certificates`, `quic`, and `protocol` are needed to handle session ticket updates. They are passed to the state via `KernelContext` (which I've just renamed from `ExternalContext`)
- `protocol_version` is used for some error checks
- `suite` is there purely to be exposed to the user. It is needed for kTLS and is also the only way to get at the `confidentiality_limit`, which I'd like to make sure is always available.

swlynch99 commented:
Pinging on this. I think I have addressed and/or responded to all the comments, so I'm leaving this comment mostly to mark that I consider this PR to be waiting on a reviewer.

ctz commented on rustls/src/conn/kernel.rs:
```diff
@@ -0,0 +1,273 @@
+    /// This method will return an error if:
+    /// - This connection is not a TLS 1.3 connection (in TLS 1.2 session tickets
+    ///   are sent as part of the handshake).
+    /// - This connection is a server connection.
```
nit: since this is defined on `KernelConnection<ClientConnectionData>` I don't think `KernelConnection<ServerConnectionData>` gets to return an error.

swlynch99 replied:
Good catch! That was left over from when I made `KernelConnection` generic. I must have missed it when making changes.

ctz commented on rustls/src/conn/kernel.rs:
```diff
@@ -0,0 +1,273 @@
+            ));
+        }
+
+        let mut reader = Reader::init(payload);
+        let nst = NewSessionTicketPayloadTls13::read(&mut reader)?;
+        reader.expect_empty("NewSessionTicket")?;
```
```suggestion
        let nst = NewSessionTicketPayloadTls13::read_bytes(payload)?;
```

ctz commented on rustls/src/conn/kernel.rs:
```diff
@@ -218,7 +218,7 @@ impl KernelConnection<ClientConnectionData> {
     pub fn handle_new_session_ticket(&mut self, payload: &[u8]) -> Result<(), Error> {
         // We want to return a more specific error here first if this is called
         // on a non-TLS 1.3 connection since a parsing error isn't the real issue
-        // here. 
```
Please squash these whitespace fixes into previous commit. (ditto below)

swlynch99 commented:
I think that's everything resolved. I have also rebased on `main`, so everything should be up to date.

djc commented on rustls/src/tls13/key_schedule.rs:
```diff
@@ -578,26 +578,30 @@ impl KeyScheduleTraffic {
-                hkdf_expand_label(expander.as_ref(), b"iv", &[]),
-            )
-        }
+    pub(crate) fn refresh_traffic_secret(
```
I think this rename means this commit won't compile, can you fix that?

swlynch99 replied:
This was not a rename (if you look at the diff you'll notice `extract_secrets` still exists below). However, `refresh_traffic_secrets` is part of the wrong commit here, so I've moved the diff adding it to [99b1da4](https://github.com/rustls/rustls/pull/2370/commits/99b1da43204855335856f358668fd2276e526aa4).

djc reviewed (approved):
Thanks for all your work on this!
